### PR TITLE
Extended tasks endpoints

### PR DESCRIPTION
**Pull request by Samridhi-98**
**Description :** Added extended features for tasks endpoint

**Fixes :** #58  

**Task List :**

- [x] Added endpoint `/tasks/index` 

- [x] Added endpoint `/tasks/index?count=3`

- [x] Added endpoint `/tasks?count=2`

### CODE REVIEW

**Comment by Samridhi-98** [comment]
@paramsiddharth  I have extended the first part i.e index for the endpoint /tasks.

File: src/api/routes/tasks.ts [hunk 1]
```diff
@@ -17,4 +17,24 @@ router.get("/", (req: Request, res: Response, next: NextFunction) => {
...
 });
 
+//GET request /tasks/index
+router.get("/:index", (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const idx = req.params.index;
+    console.log("index: ", idx)
+    console.log("data: ", data.data.length)
+    if (data.data.length > idx) {
+      res.status(200).json(data.data[idx]);
```
Comment (paramsiddharth) [inline src/api/routes/tasks.ts:27]: That's not it, @Samridhi-98 ! As I mentioned previously, the response format should be consistent. The task object must be returned as the `data` attribute of an object.
```js
{
	"data": data.data[idx]
}
```
Hence, what you should be doing is:
```js
res.status(200).json({ data: data.data[idx] });
```
Whether an array (for `/` or `?count`) or a single value `/:index`, you need to make sure the response will be an object with the data as its `data` property in JSON.

**Review by paramsiddharth — CHANGES_REQUESTED** [review]

File: src/api/routes/tasks.ts (hunk 1 above)
Comment (Samridhi-98) [reply src/api/routes/tasks.ts:27]: Okay, will update it.

File: src/api/routes/tasks.ts [hunk 2]
```diff
@@ -17,4 +17,25 @@ router.get("/", (req: Request, res: Response, next: NextFunction) => {
...
+router.get("/:index", (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const idx = req.params.index;
+    const count = req.query.count;
+    console.log("index: ", idx)
+    console.log("count: ", count);
+    if (data.data.length > idx) {
+      res.status(200).json({ data: data.data[idx] });
+    } else {
+      res.status(400).json({ message: "No data found at the given index" })
```
Comment (paramsiddharth) [inline src/api/routes/tasks.ts:30]: The status code for data not found should be 404.

**Review by paramsiddharth — CHANGES_REQUESTED** [review]

**Comment by paramsiddharth** [comment]
Excellent. Now you may continue to the next 2 tasks.

**Comment by Samridhi-98** [comment]
> Excellent. Now you may continue to the next 2 tasks.

Okay.

**Comment by Samridhi-98** [comment]
@paramsiddharth Please check I have added the count part for the tasks endpoint.

**Comment by paramsiddharth** [comment]
> @paramsiddharth Please check I have added the count part for the tasks endpoint.

Do you mean the `/tasks/:index?count=x` endpoint? I saw that, well done. Go ahead with the last task.

**Comment by Samridhi-98** [comment]
@paramsiddharth  Please review my PR, I have added all the endpoints.

**Comment by paramsiddharth** [comment]
> @paramsiddharth Please review my PR, I have added all the endpoints.

@Samridhi-98 Good job! There is one problem though.

File: src/api/routes/tasks.ts [hunk 3]
```diff
@@ -6,15 +6,55 @@ const router = express.Router();
...
   try {
-    console.log(data);
-    res.status(200).json(data);
+    let count: any = req.query.count;
+    // console.log("/-count: ", count);
+    if (count) {
+      let result = [];
+      for (let i: any = 0; i < parseInt(count); i++) {
+        result.push(data.data[i]);
+      }
```
Comment (paramsiddharth) [inline src/api/routes/tasks.ts:15]: What if `count` is greater than `data.data.length`? You need to make sure `i` doesn't exceed the maximum size of the array.

File: src/api/routes/tasks.ts [hunk 4]
```diff
@@ -6,15 +6,55 @@ const router = express.Router();
...
+router.get("/:index", (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const idx: number = parseInt(req.params.index);
+    let count: any = req.query.count;
+    // console.log("index: ", idx)
+    // console.log("count: ", count);
+    if (data.data.length > idx) {
+      if (count) {
+        let result = [];
+        const size: number = idx + parseInt(count);
```
Comment (paramsiddharth) [inline src/api/routes/tasks.ts:39]: Watch out for the same problem as above!

**Review by paramsiddharth — CHANGES_REQUESTED** [review]

**Comment by Samridhi-98** [comment]
> > @paramsiddharth Please review my PR, I have added all the endpoints.
> 
> @Samridhi-98 Good job! There is one problem though.

Is there anything that I need to change or improve?

**Comment by paramsiddharth** [comment]
> Is there anything that I need to change or improve?

Yes. Look at the changes I have requested.

File: src/api/routes/tasks.ts (hunk 4 above)
Comment (paramsiddharth) [reply src/api/routes/tasks.ts:39]: > > Watch out for the same problem as above!
> 
> In this case, it's returning the data with a possible valid result and filled the rest with null.
> Do I need to return an error message if the count is greater or anything else?

No. If the count exceeds the size, send back the data till the end only.

File: src/api/routes/tasks.ts (hunk 3 above)
Comment (paramsiddharth) [reply src/api/routes/tasks.ts:15]: Do the same as I mentioned below. That would be sufficient. You don't need to throw an error.

File: src/api/routes/tasks.ts (hunk 4 above)
Comment (Samridhi-98) [reply src/api/routes/tasks.ts:39]: Okay, I will do so.

**Comment by Samridhi-98** [comment]
@paramsiddharth  Please review my PR, fixed the mentioned changes.

**Comment by paramsiddharth** [comment]
Excellent. Thank you for contributing!

**Pull request merged**